### PR TITLE
changefeedccl: Implement kafka feed for tests.

### PR DESCRIPTION
**Pull request by miretskiy**
Make it possible to write light weight unit tests targetting
changefeeds emitting to kafka by implementing a kafka specific
TestFeed implementation.

Significantly extend test coverage by enabling majority of the
existing changefeed tests to use newly added `kafkaTest`.

Release Notes: None

### CODE REVIEW

**Review by stevendanna — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/64161)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @miretskiy)

*[a discussion](https://reviewable.io/reviews/cockroachdb/cockroach/64161#-MZCWn1Z31igMM5HudG7:-MZCWn1Z31igMM5HudG8:b-p53odq) (no related file):*
I like a lot of the cleanups here and I think having the ability to run more of our tests using the kafka sink is going to be a nice win for us.

Some of the smaller cleanups are nice:

- Moving the schema registry out into its own file and making it public
- The seenMarker
- Addition of a kafkaClient interface

One thing I think would be good is a bit more documentation around which tests aren't working currently and why.

The flow of logic between the Dial knob, kafkaFeedState, the kafkaFeedFactory, and friends can be a bit hard to follow, but I don't have a better suggestion at the moment.  

Thanks for working on this! 

---
___
*[pkg/ccl/changefeedccl/changefeed_test.go, line 3179 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/64161#-MZCVhZmEAxWftZdsw9f:-MZCVhZn5G8PEfTdSrS3:bu3jt62) ([raw file](https://github.com/cockroachdb/cockroach/blob/7c084d74562923b3ec463a04b8de66298830d013/pkg/ccl/changefeedccl/changefeed_test.go#L3179)):*
> ```Go
> 	t.Run(`enterprise-two-feeds`, enterpriseTest(memLimitTest(2)))
> 	t.Run(`enterprise-many-feeds`, enterpriseTest(memLimitTest(3)))
> 	// TODO(yevgeniy): Fix for kafka
> ```

Could we add a few words about what needs to be fixed here?
___
*[pkg/ccl/changefeedccl/helpers_test.go, line 530 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/64161#-MZCSUrw0os7ccsDzRNS:-MZCSUrw0os7ccsDzRNT:b-4p6oej) ([raw file](https://github.com/cockroachdb/cockroach/blob/7c084d74562923b3ec463a04b8de66298830d013/pkg/ccl/changefeedccl/helpers_test.go#L530)):*
> ```Go
> 		ctx := context.Background()
> 
> 		// There can be only 1 active feed.  We don't support parallel tests.
> ```

Just for clarity, it looks like we don't support parallel tests _against a single test server_ because of how the Dial testing knob and feed creation works?
___
*[pkg/ccl/changefeedccl/cdctest/avro.go, line 1 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/64161#-MZCKw4p1cS6EhiF30Mq:-MZCKw4q64dMVRGIzknP:b4gqm3r) ([raw file](https://github.com/cockroachdb/cockroach/blob/7c084d74562923b3ec463a04b8de66298830d013/pkg/ccl/changefeedccl/cdctest/avro.go#L1)):*
> ```Go
> // Copyright 2021 The Cockroach Authors.
> ```

nit: Perhaps schema_registry.go would be a better name for this file.

**Review by miretskiy — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/64161)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @stevendanna)

*[a discussion](https://reviewable.io/reviews/cockroachdb/cockroach/64161#-MZCWn1Z31igMM5HudG7:-MZDAzDo-j2vzAjke3UX:bumkbtj) (no related file):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

I like a lot of the cleanups here and I think having the ability to run more of our tests using the kafka sink is going to be a nice win for us.

Some of the smaller cleanups are nice:

- Moving the schema registry out into its own file and making it public
- The seenMarker
- Addition of a kafkaClient interface

One thing I think would be good is a bit more documentation around which tests aren't working currently and why.

The flow of logic between the Dial knob, kafkaFeedState, the kafkaFeedFactory, and friends can be a bit hard to follow, but I don't have a better suggestion at the moment.  

Thanks for working on this! 
</blockquote></details>

Thanks.  I think you're being too kind re knobs & feeds.  The whole thing is awfully complicated; very hard to follow, and very hard to debug.  I have even bigger issues with the way things are implemented: none of the feeds currently could work without the knobs setup done in helpers test ; however the feeds themselves are in a `cdctest` package while the setup is in the `changefeedccl`.  These two are entangled very "subtly" through knobs -- but I think this subtlety is not great -- I'd rather have access to internals, but be explicit as to what I'm doing, as opposed on relying on knobs.  There are numerous other things I want to fix -- alas, for this round, I wanted to focus on getting functionality down.

---
___
*[pkg/ccl/changefeedccl/changefeed_test.go, line 3179 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/64161#-MZCVhZmEAxWftZdsw9f:-MZDCDV4-y3_cDBtWh-J:b-63qzuz) ([raw file](https://github.com/cockroachdb/cockroach/blob/7c084d74562923b3ec463a04b8de66298830d013/pkg/ccl/changefeedccl/changefeed_test.go#L3179)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

Could we add a few words about what needs to be fixed here?
</blockquote></details>

I added comment in the `kafkaTest` function -- see around knobs section.
___
*[pkg/ccl/changefeedccl/helpers_test.go, line 530 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/64161#-MZCSUrw0os7ccsDzRNS:-MZDC8viDLnQGB01I2KB:bwy8pfz) ([raw file](https://github.com/cockroachdb/cockroach/blob/7c084d74562923b3ec463a04b8de66298830d013/pkg/ccl/changefeedccl/helpers_test.go#L530)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

Just for clarity, it looks like we don't support parallel tests _against a single test server_ because of how the Dial testing knob and feed creation works?
</blockquote></details>

Yes -- correct.  This whole business of knobs is what's stopping this.
___
*[pkg/ccl/changefeedccl/cdctest/avro.go, line 1 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/64161#-MZCKw4p1cS6EhiF30Mq:-MZDDdeFE0A0SXAs-rrT:b-fhkj3n) ([raw file](https://github.com/cockroachdb/cockroach/blob/7c084d74562923b3ec463a04b8de66298830d013/pkg/ccl/changefeedccl/cdctest/avro.go#L1)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

nit: Perhaps schema_registry.go would be a better name for this file. 
</blockquote></details>

Sure -- initially, I thought I'd have to move more code into cdctest package.

File: pkg/ccl/changefeedccl/helpers_test.go [hunk 1]
```diff
@@ -396,6 +382,186 @@ func cloudStorageTest(
...
+
+func watchdog(timeout time.Duration) (cancel func(), wait func()) {
+	var wg sync.WaitGroup
+	wg.Add(1)
+	cancelCh := make(chan struct{})
+	go func() {
+		defer wg.Done()
+		select {
+		case <-time.After(timeout):
+			panic(fmt.Sprintf("test timeout\n%s", debug.Stack()))
```
Comment (stevendanna) [inline pkg/ccl/changefeedccl/helpers_test.go:457]: I'm a little worried about this because I know we had a recent issue with a panic in a goroutine stopping the entire test runner meaning that the nightlies had a lot of tests that weren't actually being run.

**Review by stevendanna — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/64161)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner, @miretskiy, and @stevendanna)

*[a discussion](https://reviewable.io/reviews/cockroachdb/cockroach/64161#-MZVDCbPCgPCV6YvG2dx:-MZVDCbPCgPCV6YvG2dy:b-lrt7yq) (no related file):*
:lgtm: Thanks for getting this going!

---

**Comment by miretskiy** [comment]
bors r=ssd
